clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java

```diff
@@ -71,12 +71,12 @@
     /**
      * @see KafkaConsumer#commit(Map, CommitType)
      */
-    public void commit(Map<TopicPartition, Long> offsets, CommitType commitType);
+    public void commit(Map<TopicPartition, OffsetMetadata> offsets, CommitType commitType);
```

Comment:
The problem with the `List` variant is that it's inconvenient for the application to store it that way -- it's most convenient if the API uses a layout that works well for applications, and since the application probably needs to do lookups by `TopicPartition`, using a `Map` is natural. Also, the map guarantees correct semantics, i.e. only one entry per `TopicPartition`.

I think requiring `OffsetAndMetadata` or whatever name you want to use is fine as long as it's mutable. It's a little overhead for people who only want to set the offset manually, but it's tiny and if its mutable there's not much to complain about wrt object allocation overhead or anything like that.

Also, I realized that the ownership requirements for the data associated with async commits is probably not clear from the javadocs. i.e. do I need to make sure the data I passed in doesn't change until the commit callback is invoked, or is KafkaConsumer going to ensure the data is copied so I'm free to start modifying the map I pass in as soon as the commit method returns?